compiler/rustc_codegen_cranelift/src/value_and_place.rs

```diff
@@ -231,22 +231,24 @@ impl<'tcx> CValue<'tcx> {
     pub(crate) fn const_val(
         fx: &mut FunctionCx<'_, 'tcx, impl Module>,
         layout: TyAndLayout<'tcx>,
-        const_val: u128,
+        const_val: ty::ScalarInt,
     ) -> CValue<'tcx> {
+        assert_eq!(const_val.size(), layout.size);
         use cranelift_codegen::ir::immediates::{Ieee32, Ieee64};
 
         let clif_ty = fx.clif_type(layout.ty).unwrap();
 
         if let ty::Bool = layout.ty.kind() {
             assert!(
-                const_val == 0 || const_val == 1,
+                const_val == ty::ScalarInt::FALSE || const_val == ty::ScalarInt::TRUE,
                 "Invalid bool 0x{:032X}",
                 const_val
             );
         }
 
         let val = match layout.ty.kind() {
             ty::Uint(UintTy::U128) | ty::Int(IntTy::I128) => {
+                let const_val = const_val.to_bits(layout.size).unwrap();
```

Comment:
Below `u64::try_from(const_val)` is used. This is independent of the expected size of the value. For `ty::ScalarInt` it will fail if it isn't 64bits. That code requires the value to be zero-extended to 64bits instead of failing when it is smaller..

Reply:
I also added the size checked conversion to the code below

Reply:
I think it will work fine now.